[DEV] Add service `geonature-atlas`

VincentCauchois opened the pull request:
# OBJECTIF

🎯 Permettre l'ajout d'un service `geonature-atlas` à une stack GDS, i.e. l'ajout de l'application GeoNature-Atlas à une instance GN déployée avec GDS.

🎯 Cibler l'ajout de GeoNature-Atlas en version 1.7.0 avec GeoNature en version 2.16.1


# CONTEXTE

🧱 Reprise du travail initié dans https://github.com/PnX-SI/GeoNature-Docker-services/pull/51 par @jpm-cbna

▶️ Rebase de ce travail sur `develop` à jour, avec de nombreuses évolutions de GDS dont l'amélioration du mode dev ainsi que la possibilité d'utiliser plusieurs fichiers `.yml` pour le déploiement d'une stack et la possibilité de sélectionner _à la carte_ les services souhaités en utilisant les _profiles_.

⚠️ Avec la version 2 de GeoNature-Atlas - travail en cours sur la branche `develop` du dépôt - les scripts d'installation `install_db.sh` et `install_app.sh` devraient évoluer pour être utilisables directement lors de l'instanciation d'une stack GDS.

📌 En attendant cette version 2, on continuera d'utiliser le script `docker_install_atlas_schema.sh`, script qui devra probablement être modifié afin de le rendre compatible avec la version 1.7.0 de GeoNature-Atlas.


# TODO

### Reprise du travail fait par @jpm-cbna 
- [x] Cherry-pick le commit de la PR https://github.com/PnX-SI/GeoNature-Docker-services/pull/51 avec premières adaptations
- [x] Faire un premier essai de déploiement d'une stack GDS avec le nouveau service `geonature-atlas`
- [x] Éventuellement, adapter jusqu'à ce que le déploiement avec `geonature-atlas` fonctionne // spécificités de GDS sur `feat/docker-evolution` et de GN-A en 1.7.0
- [ ] Prise en compte des autres développements de @jpm-cbna sur son fork de GDS : 
  - [ ] `.env.sample` : https://github.com/cbn-alpin/atlas-docker/blob/main/.env.sample
  - [ ] `docker-compose.yml` : https://github.com/cbn-alpin/atlas-docker/blob/main/docker-compose.yml
  - [ ] Patch du `docker-compose.yml` : https://github.com/cbn-alpin/gentiana-srv/blob/main/geonature/docker-compose.yml.patch
  - [ ] `docker-compose.override.yml` : https://github.com/cbn-alpin/gentiana-srv/blob/main/geonature/docker-compose.override.yml
  - [x] Dernières évolutions sur branche `feat/geonature-atlas` : https://github.com/cbn-alpin/gentiana-srv/compare/main...feat/geonature-atlas

### Refactorisations et améliorations
- [ ] Améliorations en cohérence aves les évolutions de `feat/docker-evolution` :
  - [ ] Permettre l'utilisation avec Traefik
  - [x] Utiliser un profile `atlas` dédié pour le service `geonature-atlas` et ne pas utiliser ce profile par défaut
  - [ ] Adapter le script `generate_env.py`
  - [ ] Garantir le bon fonctionnement et en mode `prod` et en mode `dev`
    - [ ] Adapter le fichier `dev.yml`
    - [ ] ...
  - [ ] Compléter la documentation : "README.md" et/ou "FAQ"
- [ ] Récupération des améliorations de JP sur le dépôt GNA :
  - [ ] Basculer sur la branche `feat/release-1.7.1` ou la version 1.7.1 de GeoNature-Atlas - avec correctifs des scripts d'installation Docker de l'Atlas - de Jean-Pascal
    - 📓 //  Testée par JP avec un `docker-compose.yml` qui s'appuie sur une base GN locale à sa machine hôte // dépôt dédié : https://github.com/cbn-alpin/atlas-docker
    - 📓 // Ticket lié :  https://github.com/PnX-SI/GeoNature-atlas/issues/660
  - [ ] Récupération du commit de JP https://github.com/PnX-SI/GeoNature-atlas/commit/6e7c348cdbeed58fbd33a61088363438f73cdf3c d'optimisation de la génération de `vm_observations`
  - [ ] Vérifications du TODO de la PR initiale : https://github.com/PnX-SI/GeoNature-Docker-services/pull/51#issuecomment-2602397824
    - [ ] (?) Gestion de l'initialisation du service `geonature-atlas` après l'initialisation avec succès des services `postgres` et `geonature-install-db` : s'inspirer de `usershub_entrypoint.sh` https://github.com/PnX-SI/GeoNature-Docker-services/pull/51#issuecomment-2602397824 ?
    - [ ] (?) Régler le problème de l'interpolation des variables PSQL lors de l'installation de l'Atlas
    - [ ] (?) Trancher quant à l'utilisation des 2 variable `ATLAS_INSTALL_SCHEMA` et `ATLAS_RESET_SCHEMA`
    - [ ] (?) Gestion via secrets pour variables d'environnements // voir ce qui a été fait sur `feat/docker-evolution`

jpm-cbna committed:
feat(atlas): add service `geonature-atlas`

From cherry-picked and adapted commit https://github.com/PnX-SI/GeoNature-Docker-services/pull/51/commits/2a2b5ab9929a01b988c25ac4baf644d91d3b40cc

Co-authored-by: Jean-Pascal MILCENT <jp.milcent@cbn-alpin.fr>
Co-authored-by: VincentCauchois <vincent.cauchois@mnhn.fr>

christophe-ramet commented on assets/atlas/healthcheck.sh:
à reflechir

Voir comment sont fait les autres healtcheck. Envisager installation de dockerized dans le docker ?

christophe-ramet commented on build/atlas/Dockerfile:
Pour l'instant pas utilisé. Plutôt à fix dans l'atlas

christophe-ramet commented on build/postgres/Dockerfile:
à sortir de GDS

VincentCauchois committed:
[WIP]For several commits

- .env.sample + essential.yml: Add profile 'atlas'
- essential.yml: Add ATLAS_TAXHUB_DISPLAYED_ATTRIBUTES env variable --> to go with first commit // JP
- dev.yml, Dockerfile: Add DEV mode for Atlas --> to go with [.gitmodules, GeoNature-atlas/, dev.yml] from first commit
  - Dockerfile should rather be put in ./build/dev/atlas/

christophe-ramet commented on init-config.sh:
```diff
@@ -22,6 +22,11 @@ if [ ! -f "$GEONATURE_CONFIG_DIR/usershub/config.py" ]; then
 fi
 
+if [ ! -f "./sources/GeoNature-atlas/atlas/configuration/config.py" ]; then
+    mkdir -p "./sources/GeoNature-atlas/atlas/configuration/config.py"
```
```suggestion
    mkdir -p "./sources/GeoNature-atlas/atlas/configuration/"

```

christophe-ramet commented on README.md:
```diff
@@ -1,30 +1,16 @@
+# GeoNature Docker Services
 
-Ce dépôt permet de déployer automatiquement et facilement GeoNature, UsersHub dans un environnement dockerisé et accessible en HTTPS.
+Ce dépôt permet de déployer automatiquement et facilement GeoNature, UsersHub et GeoNature-atlas - optionnellement - dans un environnement dockerisé et accessible en HTTPS.
```
Je pense pas qu'on précise ici vu que c'est optionnel et qu'on est dans la ligne de présentation du repo.

christophe-ramet commented on README.md:
```diff
@@ -52,12 +38,40 @@ De plus, ce dépôt fournit une image Docker de GeoNature contenant les modules
    - `ACME_EMAIL`: Adresse mail utilisée pour la génération du certificat SSL par [Let's Encrypt](https://letsencrypt.org/fr/) via traefik
 
-8. **Initialisation des fichiers de configurations.** Lancer la commande `./init-config.sh` afin de créer les dossiers et les fichiers de configuration requis. Le script `init-config.sh` génère aléatoirement aussi les clés secrètes pour GeoNature et UsersHub respectivement dans les fichiers suivants :
+8. Si GeoNature-atlas est activé, lancer la commande suivante :
```
Je vois le souci qui est que l'on doit configurer Atlas avec config.py, settings.ini.sample et custom. 

Premièrement, je pense qu'il faudrait éviter d'avoir à faire un `make submodule_init`, c'est pas normal que l'on ai besoin du code source de l'atlas pour lancer son docker. 

Deuxièmement, c'est pas fou qu'on doive remplir soi-même config, settings et custom avant de lancer l'atlas. Pour les autres services, on peut lancer sans avoir à faire des actions manuelles de ce genre. En plus, on se retrouve à avoir des informations dupliquées. Par exemple, `SQLALCHEMY_DATABASE_URI` de config.py est composé de variables d'environnement disponibles dans le .env ,même chose avec `URL_APPLICATION` de settings.ini. 

Idéalement il faudrait que ce genre d'information ne soit pas obligatoire dans l'atlas, mais qu'à l'image de GeoNature, on puisse les passer en variables d'environnement et ne pas les remplir dans config.py.

En attendant, on peut peut être dans init-config.sh: 

- Si config.py et settings.ini sont pas existants: 
– Récupérer config.py.sample et settings.ini.sample avec un wget;
- les copier vers config.py et settings.ini
- Les modifier à base de sed dans init-config.sh pour utiliser les variables du .env

TheoLechemia replied:
Dans le .env, je rajouterai un paramètre `GEONATURE_ATLAS_DB_NAME` pour bien forcer que la bdd soit différente que celle de GN. Et comme on s'est dit avec Vincent, copiez les variable du .env dans settings.ini via sed.
C'est vrai que c'est chiant d'avoir a télécharger le code source, mais le fichier `settings.ini` est vraiment obligatoire pour l'installation de la base (le .env ne suffit pas). Du coup il faut à minima ce fichier (peut être télécharger que ce fichier dans init-config.sh ?

christophe-ramet commented on README.md:
````diff
@@ -52,12 +38,40 @@ De plus, ce dépôt fournit une image Docker de GeoNature contenant les modules
+   ```
+   - N.B. : pour une mise à jour ultérieure des données, voir la section "## Comment mettre à jour les données de GeoNature-atlas ?" de la FÀQ
+
+13. Si GeoNature-atlas est activé, il faut, au premier lancement, peupler le dossier "sources/GeoNature-atlas/atlas/static/custom" - on peut partir de ce qui est dans "sources/GeoNature-atlas/atlas/static/custom_samples" - avec la commande suivante, et, dans l'idéal, adapter :
````
Idem, c'est pas fou qu'on doive copier le custom_sample. 

Idéalement, l'atlas fait comme dans Geonature et si un fichier de custom n'est pas spécifié, il utilise le custom_sample. 

En attendant, on peut faire dans un nouvel entrypoint :

- Copier custom_sample vers custom seulement pour les fichiers n'existant pas déjà dans custom (à tester, mais :
`cp -rn source/ destination/`)

TheoLechemia replied:
la copie des fichier custom est faite par `install_app.sh` qui est lancer dans le DockerFile
https://github.com/PnX-SI/GeoNature-atlas/blob/master/Dockerfile#L62

christophe-ramet commented on docker-compose.essential.yml:
```diff
@@ -171,6 +171,23 @@ services:
+      postgres:
+        condition: service_healthy
+    volumes:
+      - ./sources/GeoNature-atlas/atlas/configuration/config.py:/dist/atlas/configuration/config.py
+      - ./sources/GeoNature-atlas/atlas/configuration/settings.ini:/dist/atlas/configuration/settings.ini
+      - ./sources/GeoNature-atlas/atlas/static/custom:/dist/atlas/static/custom
```
```suggestion
      - ${GEONATURE_DATA_DIR:-./data}/atlas/configuration/config.py:/dist/atlas/configuration/config.py
      - ${GEONATURE_DATA_DIR:-./data}/atlas/configuration/settings.ini:/dist/atlas/configuration/settings.ini
      - ${GEONATURE_DATA_DIR:-./data}//atlas/static/custom:/dist/atlas/static/custom
```

Comme ça on est coherent avec geonature et on evite de modifier dans le sous module

christophe-ramet commented on docker-compose.traefik.yml:
```diff
@@ -71,3 +71,11 @@ services:
       - gn
+
+  geonature-atlas:
+    labels:
+      - "traefik.enable=true"
+      - "traefik.http.routers.geonature-atlas.rule=Host(`${ATLAS_HOST}`) && PathPrefix(`${ATLAS_PREFIX}`)"
+      - "traefik.http.routers.geonature-atlas.entrypoints=websecure"
+      - "traefik.http.routers.geonature-atlas.tls.certResolver=acme-resolver"
```
```suggestion
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.geonature-atlas.rule=Host(`${ATLAS_HOST}`) && PathPrefix(`${ATLAS_PREFIX}`)"
      - "traefik.http.routers.geonature-atlas.entrypoints=websecure"
      - "traefik.http.routers.geonature-atlas.tls.certResolver=acme-resolver"
       - "traefik.http.services.geonature-atlas.loadbalancer.server.port=8080"
```
Pour être explicite et coherent avec les autres services

TheoLechemia commented on build/dev/atlas/entrypoint-dev.sh:
```diff
@@ -0,0 +1,5 @@
+
+cd /dist/atlas/static && npm ci --omit=dev
+
+gunicorn -b 0.0.0.0:8080 "atlas.app:create_app()" --reload --reload-extra-file="/dist/atlas/configuration/config.py"
```
L'image Docker de GN-Atlas propose déjà un stage de dev 
-> 
https://github.com/PnX-SI/GeoNature-atlas/blob/master/Dockerfile#L130
En modifiant le docker compose pour utiliser ce stage on peut se passer de cet entrypoint

TheoLechemia commented on docker-compose.dev.yml:
```diff
@@ -112,4 +108,14 @@ services:
+      - ./build/dev/atlas/entrypoint-dev.sh:/entrypoint-dev.sh
+    build:
+      context: sources/GeoNature-atlas
+      target: dev
```
niquel, le target  dev est déja utilisé, donc on peut virer l'entrypoint
